watertap/unit_models/selective_oil_permeation.py

```diff
@@ -0,0 +1,669 @@
+
+        self.den_constant = Param(
+            mutable=True,
+            initialize=2.66e15,
+            units=units.dimensionless,
+            doc="Denominator constant",
+        )
+
+        self.den_mu_exp = Param(
+            mutable=True,
+            initialize=1.1,
+            units=units.dimensionless,
+            doc="Denominator viscosity exponent",
+        )
+
+        self.den_PT_exp = Param(
+            mutable=True,
+            initialize=-2.1,
+            units=units.dimensionless,
+            doc="Denominator transmembrane pressure exponent",
+        )
+
+        self.den_v_exp = Param(
+            mutable=True,
+            initialize=0.4,
+            units=units.dimensionless,
+            doc="Denominator liquid velocity exponent",
+        )
+
+        # Add unit variables
+        self.flux_vol_oil = Var(
+            self.flowsheet().config.time,
+            initialize=1e-8,
+            bounds=(0.0, None),
+            units=units_meta("length") * units_meta("time") ** -1,
+            doc="Oil volumetric flux",
+        )
+
+        self.flux_vol_oil_pure = Var(
+            self.flowsheet().config.time,
+            initialize=1e-8,
+            bounds=(0.0, None),
+            units=units_meta("length") * units_meta("time") ** -1,
+            doc="Oil volumetric flux if the feed was pure oil",
+        )
+
+        self.effective_area_ratio = Var(
+            self.flowsheet().config.time,
+            initialize=0.1,
+            bounds=(0.0, 1.0),
+            units=units.dimensionless,
+            doc="Effective area ratio",
+        )
+
+        self.recovery_frac_oil = Var(
+            self.flowsheet().config.time,
+            initialize=0.5,
+            bounds=(0, 1),
+            units=units.dimensionless,
+            doc="Oil recovery fraction on a mass basis",
+        )
+
+        self.mass_transfer_oil = Var(
+            self.flowsheet().config.time,
+            initialize=1e-6,
+            bounds=(0.0, None),
+            domain=NonNegativeReals,
+            units=units_meta("mass") * units_meta("time") ** -1,
+            doc="Mass transfer of oil to permeate",
+        )
+
+        self.area = Var(
+            initialize=1,
+            bounds=(0.0, 1e6),
+            domain=NonNegativeReals,
+            units=units_meta("length") ** 2,
+            doc="Membrane area",
+        )
+
+        self.pressure_transmemb = Var(
+            self.flowsheet().config.time,
+            initialize=1e5,
+            bounds=(0.0, 1e8),
+            domain=NonNegativeReals,
+            units=units_meta("pressure"),
+            doc="Average transmembrane pressure",
+        )
+
+        self.liquid_velocity = Var(
+            self.flowsheet().config.time,
+            initialize=0.01,
+            bounds=(0.0, 1e3),
+            domain=NonNegativeReals,
+            units=units_meta("length") * units_meta("time") ** -1,
+            doc="Shell side liquid velocity at module inlet",
```

Comment:
I'd think that liquid_velocity could be calculated at the inlet and outlet, but this is set for inlet only. Maybe expand to include inlet and outlet?

Reply:
For this model, the liquid velocity only at the inlet is needed - inlet liquid velocity is used in the correlation for effective area ratio. We could similarly calculate the outlet liquid velocity, but it's not needed for anything, so I haven't bothered to implement it.